src/olympia/files/models.py

```diff
@@ -519,47 +520,45 @@ def save(self, *args, **kw):
+            for chunk in chunks:
+                hash_obj.update(chunk)
+                file_destination.write(chunk)
+        return hash_obj
+
     def add_file(self, chunks, filename, size):
         if not self.uuid:
             self.uuid = self._meta.get_field('uuid')._create_uuid()
 
-        filename = force_text(u'{0}_{1}'.format(self.uuid.hex, filename))
-        loc = os.path.join(user_media_path('addons'), 'temp', uuid.uuid4().hex)
-        base, ext = os.path.splitext(filename)
-        is_crx = False
-
-        # Change a ZIP to an XPI, to maintain backward compatibility
-        # with older versions of Firefox and to keep the rest of the XPI code
-        # path as consistent as possible for ZIP uploads.
-        # See: https://github.com/mozilla/addons-server/pull/2785
-        if ext == '.zip':
-            ext = '.xpi'
-
-        # If the extension is a CRX, we need to do some actual work to it
-        # before we just convert it to an XPI. We strip the header from the
-        # CRX, then it's good; see more about the CRX file format here:
-        # https://developer.chrome.com/extensions/crx
-        if ext == '.crx':
-            ext = '.xpi'
-            is_crx = True
-
-        if ext in amo.VALID_ADDON_FILE_EXTENSIONS:
-            loc += ext
-
-        log.info('UPLOAD: %r (%s bytes) to %r' % (filename, size, loc),
+        _base, ext = os.path.splitext(filename)
+        was_crx = ext == '.crx'
+        # Filename we'll expose (but not use for storage).
+        self.name = force_text('{0}_{1}'.format(self.uuid.hex, filename))
+
+        # Final path on our filesystem. Will always end in .xpi no matter what
+        # is uploaded. CRX files are converted beforehand. If somehow this is
+        # not a valid archive parse_addon() will eventually complain at
+        # validation time.
+        self.path = os.path.join(
+            user_media_path('addons'), 'temp', uuid.uuid4().hex, '.xpi')
+
+        log.info('UPLOAD: %r (%s bytes) to %r' % (self.name, size, self.path),
                  extra={'email': (self.user.email
                                   if self.user and self.user.email else '')})
-        if is_crx:
-            hash_func = write_crx_as_xpi(chunks, loc)
-        else:
-            hash_func = hashlib.sha256()
-            with storage.open(loc, 'wb') as file_destination:
-                for chunk in chunks:
-                    hash_func.update(chunk)
-                    file_destination.write(chunk)
-        self.path = loc
-        self.name = filename
-        self.hash = 'sha256:%s' % hash_func.hexdigest()
+
+        hash_obj = None
+        if was_crx:
+            try:
+                hash_obj = write_crx_as_xpi(chunks, self.path)
+            except InvalidOrUnsupportedCrx:
+                # We couldn't convert the crx file. Write it to the filesystem
+                # normally, the linter should reject this with a proper error.
```

Comment:
in the pr comments it says the error happens before file gets to the linter

Reply:
Yeah I need to fix the comment